src/honey_badger/epoch_state.rs

```diff
@@ -230,54 +263,122 @@ where
             if has_seen_done {
                 error!("`SubsetOutput::Done` was not the last `SubsetOutput`");
             }
-            match cs_output {
-                SubsetOutput::Contribution(k, v) => {
-                    step.extend(self.send_decryption_share(k.clone(), &v)?);
-                    self.accepted_proposers.insert(k);
-                }
-                SubsetOutput::Done => {
-                    self.subset = SubsetState::Complete(self.accepted_proposers.clone());
-
-                    let faulty_shares: Vec<_> = self
-                        .decryption
-                        .keys()
-                        .filter(|id| !self.accepted_proposers.contains(id))
-                        .cloned()
-                        .collect();
-                    for id in faulty_shares {
-                        if let Some(DecryptionState::Ongoing(td)) = self.decryption.remove(&id) {
-                            for id in td.sender_ids() {
-                                let fault_kind = FaultKind::UnexpectedDecryptionShare;
-                                step.fault_log.append(id.clone(), fault_kind);
-                            }
-                        }
+
+            let Self {
+                accepted_proposers,
+                epoch,
+                decryption,
+                netinfo,
+                subset,
+                subset_handler,
+                ..
+            } = self;
+
+            macro_rules! send_decryption_share {
+                ($proposer_id: expr, $data: expr) => {
+                    step.extend(Self::send_decryption_share(
+                        *epoch,
+                        decryption,
+                        netinfo.clone(),
+                        $proposer_id.clone(),
+                        &$data,
+                    )?);
+                };
+            }
+
+            macro_rules! finalize_subset_with_step {
+                () => {{
+                    has_seen_done = true;
+                    Self::finalize_subset_with_step(subset, accepted_proposers, decryption, &mut step);
+                }};
+            }
```

Comment:
It's a matter of taste, but I generally prefer using macros sparingly. I find it harder to read, and I think tools like Rustfmt and Clippy do, too.

Reply:
The motivation here was to avoid typing out the (somewhat long) list of arguments here -- a closure would have resulted in ownership problems, and I wanted to factor out this call specifically.

That said, I agree that rolling macros in this way is definitely harder on maintenance. Do you think it'd just be better to duplicate the calls?

Reply:
Doesn't a macro compile to a call every time you call it? It's better to call the intended function instead of hiding it behind an obscure macro.

Reply:
@vkomenda: Not sure what you mean here -- Rust will compile the content of this particular call to a function call, since that's the content of the macro, but it won't create another layer of calling a function (which I'm not sure you meant, but it sounded like). It's basically a copy-paste in any context you use it. Point taken, though -- I'll de-normalize the calls. :)

Again, the intent of the macro was to "bind" arguments without running afoul of `borrowck` -- if one were to use a lambda expression, for instance, then there'd be borrowing clashes with `accepted_proposers` because `accepted_proposers` would get used before this ever got used.

Reply:
I think with the approach below we should be able to have simpler code without macros and without much repetition.

Reply:
@ErichDonGubler, please correct me if I'm wrong but two occurrences of a macro that is calling a function in Rust expand to two different function calls. In any case, the bottomline is that macros are a last resort option, and there are multiple ways you can continue without introducing a macro here. These macros increase code complexity.

Reply:
@vkomenda Yep, okay, we're on the same page. Sorry for any confusion!